### PR TITLE
Add set_vcf_sample_id sample.

### PR DESCRIPTION
**Pull request by mbookman**

### CODE REVIEW

File: set_vcf_sample_id/cloud/run_set_vcf_sample_id.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+parser.add_argument("--logging", required=True,
+                    help="Cloud Storage path to send logging output")
+parser.add_argument("--poll-interval", default=0, type=int,
+                    help="Frequency (in seconds) to poll for completion (default: no polling)")
+args = parser.parse_args()
+args.script_path.rstrip('/')
+
+# Create the genomics service
+credentials = GoogleCredentials.get_application_default()
+service = build('genomics', 'v1alpha2', credentials=credentials)
```
Comment (pgrosu) [inline set_vcf_sample_id/cloud/run_set_vcf_sample_id.py]: nit, I thought this was now at `v1`?

Comment (pgrosu) [inline set_vcf_sample_id/cloud/run_set_vcf_sample_id.py]: Do the Genomics credentials relate to the the API version they provide permission to?  So if the pipeline API is `v1alpha2` does that mean the credentials have to be the same version?

~p

Comment (mbookman) [inline set_vcf_sample_id/cloud/run_set_vcf_sample_id.py]: The bulk of the genomics API is v1.
The Pipelines API is in open alpha.

Credentials are independent of API version.

Comment (pgrosu) [inline set_vcf_sample_id/cloud/run_set_vcf_sample_id.py]: Yup, that makes sense :)  So would it be correct to say that the `v1alpha2` in the script would refer to the API it gives permissions to, which in this case would be the pipelines.

Thanks,
~p

File: set_vcf_sample_id/README.md [hunk 2]
```diff
@@ -0,0 +1,185 @@
+# Update a VCF header line sample ID
+
+This pipeline is useful when you have a set of single-sample VCFs
+in Cloud Storage in which the sample ID in the header needs to be changed.
```
Comment (deflaux) [inline set_vcf_sample_id/README.md]: Please mention `gcloud alpha genomics callsets update --help` which let's you fix call_set_name _after_ an import somewhere in this readme as an aside.

File: set_vcf_sample_id/README.md [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+| This example demonstrates using the Pipelines API to run custom code *without building a Docker image*. In this example, the scripts are copied into Cloud Storage and then downloaded at pipeline run time.  The docker image used is the stock [Python 2.7 image from dockerhub](https://hub.docker.com/_/python/). |
+
+| API Note |
+|----------|
+| This example demonstrates the use of non-file input parameters. These parameters (`ORIGINAL_SAMPLE_ID` and `NEW_SAMPLE_ID`) get set in the environment and are then available to the code running in the docker container. |
+
+## (1) Copy scripts to Cloud Storage
+
+```
+gsutil cp set_vcf_sample_id.sh set_vcf_sample_id.py \
```
Comment (deflaux) [inline set_vcf_sample_id/README.md]: typo: identical filename listed twice

File: set_vcf_sample_id/cloud/run_set_vcf_sample_id.py (hunk 1 above)
Comment (mbookman) [inline set_vcf_sample_id/cloud/run_set_vcf_sample_id.py]: Here is where it [gets the credentials](https://developers.google.com/identity/protocols/application-default-credentials):
```
credentials = GoogleCredentials.get_application_default()
```

Here it says "build my client interface to the genomics v1alpha2 API if the credentials authorize access":

```
service = build('genomics', 'v1alpha2', credentials=credentials)
```

File: set_vcf_sample_id/README.md (hunk 3 above)
Comment (mbookman) [inline set_vcf_sample_id/README.md]: The first is a .sh, the second a .py.

File: set_vcf_sample_id/README.md (hunk 2 above)
Comment (pgrosu) [inline set_vcf_sample_id/README.md]: Can pipelines be run from data loaded in the Genomics API?  That was a concern of mine that the Pipeline API bypasses the Genomics API, which could benefit from more integration as all that processed data is not organized and properly searchable, which would become a bigger missed opportunity down the line.  I posted this at the following link:

https://github.com/googlegenomics/pipelines-api-examples/issues/13#issuecomment-195080153

File: set_vcf_sample_id/README.md (hunk 3 above)
Comment (deflaux) [inline set_vcf_sample_id/README.md]: I'm wrong about the typo.

But consider renaming one of these two files so that they appear to be two different files?  For example the shell script name could mention that it also takes care of compression.

File: set_vcf_sample_id/README.md [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+  gs://YOUR-BUCKET/pipelines/set_vcf_sample_id/
+```
```
Comment (deflaux) [inline set_vcf_sample_id/README.md]: This example uses subdirectory `pipelines`.

For the samtools and bioconductor samples, the pattern we've been using is:
```
gs://YOUR-BUCKET/pipelines-api-examples/<particular example name>/
gs://YOUR-BUCKET/pipelines-api-examples/<particular example name>/<scripts here, if applicable>
gs://YOUR-BUCKET/pipelines-api-examples/<particular example name>/output/<output here>
gs://YOUR-BUCKET/pipelines-api-examples/<particular example name>/logging/<logs here>
```

For fastqc:
```
  --output gs://YOUR-BUCKET/fastqc/output/<output here>
  --logging gs://YOUR-BUCKET/fastqc/logging/<logs here>
```

For compress:
```
  --output gs://YOUR-BUCKET/compress/output_path/<output here>
  --logging gs://YOUR-BUCKET/compress/logging_path/<logs here>
```

Let's pick one convention, use it for this sample, and file a bug for the other samples to be updated to the convention.

**Comment by deflaux** [comment]
left some minor feedback but otherwise LGTM!

File: set_vcf_sample_id/cloud/run_set_vcf_sample_id.py (hunk 1 above)
Comment (pgrosu) [inline set_vcf_sample_id/cloud/run_set_vcf_sample_id.py]: That helps, but when I look at the specific code for the [`build`](https://github.com/google/google-api-python-client/blob/6d7f655d572415d8d8c812f79e83a46557921a53/googleapiclient/discovery.py#L155-L218) method, it points to the version of the API, which in this case would be for `genomics`.   Thus the version might probably be `v1` instead of `v1alpha2`, based on the following code - since I'm not sure you are using the Annotation API - which would require `v1beta2` that [Nicole previously made me aware of](https://github.com/googlegenomics/dataflow-java/pull/169#issuecomment-186358503):

```Python

@positional(2)
def build(serviceName,
          version,
          ...
  """Construct a Resource for interacting with an API.

  Construct a Resource object for interacting with an API. The serviceName and
  version are the names from the Discovery service.

  Args:
    serviceName: string, name of the service.
    version: string, the version of the service.
    ...

  Returns:
    A Resource object with methods for interacting with the service.
  """

  params = {
      'api': serviceName,
      'apiVersion': version
      }
  ...
  for discovery_url in (discoveryServiceUrl, V2_DISCOVERY_URI,):
    requested_url = uritemplate.expand(discovery_url, params)

    try:
      content = _retrieve_discovery_doc(requested_url, http, cache_discovery,
                                        cache)
      return build_from_document(content, base=discovery_url, http=http,
          developerKey=developerKey, model=model, requestBuilder=requestBuilder,
          credentials=credentials)
          ...
```

Comment (mbookman) [inline set_vcf_sample_id/cloud/run_set_vcf_sample_id.py]: The call in this file simply submits the request using the Pipelines API, probably from a users workstation/laptop.

The code running on the VM can do whatever it wants - access the Genomics API, access the BigQuery API, etc. independent of how the Pipeline was submitted.

File: set_vcf_sample_id/README.md (hunk 2 above)
Comment (mbookman) [inline set_vcf_sample_id/README.md]: Updated the README to include a reference to the `callsets update` command.

Comment (mbookman) [inline set_vcf_sample_id/README.md]: Yes, the code running on the VM can access the Genomics API. The primary purpose the Pipelines API serves today is to enable people with code that uses traditional file-based data access (BAMs and VCFs) to run those at scale.

Code that uses the GA4GH API for a "pipeline" at scale already has Dataflow available, but there will probably be use-cases where either Dataflow is a big hammer or people want to use a language not supported by Dataflow.

Comment (pgrosu) [inline set_vcf_sample_id/README.md]: Agreed that Dataflow has full integration with the Genomics API, but integrating Dataflow with any language is fairly straightforward - I'm not seeing that as a bottleneck.  Yes it is Java, but you can modularize that to make it callable from other languages.

Regarding accessing the Genomics API from the Pipeline API, I'm not seeing that connection quite yet.  Do you mean that you can run a `cmd` to get data from the Genomics API, or that you have a direct integration via the Pipeline API?  Do you have any examples? 

I'm not against having the ability of the Pipeline API to provide traditional file-based data access, but there is no experimental and sample annotation of the data in the Genomics API of these files, even if these files are not loaded in the Genomics API.  Basically what will happen, is that people will begin using the Pipeline API more and more with time since it is so convenient, and not bother with the Genomics API.  Any annotation regarding the VCFs or BAMs is non-existent since they are just files, so there is no way to perform correlations among datasets.  If these files get removed, then the provenance is also gone.  And building a query system using relational algebra or even via something like PyTables is not that difficult in the cloud.  Even using something simple by implementing Sorted Strings Table (SSTable) files would allow users to bypass BigQuery.

Basically the point I am driving is that you have a nice beginning built with all these tools, especially the Genomics API.  Processing at large scale is just the beginning, as this has been done many times before.  What has not been done before is to have large scale modular integration with annotation of all types.  So there just a little more work to do, by adding the extra functionality to provide tight integration with a complete turn-key solution, or otherwise you're throwing the baby out with the bathwater, and then users keep shifting their data to the "next great" cloud-enabled data-processing platform.  Basically one then loses the reliable and expected consistency to build a fruitful user-community from.

File: set_vcf_sample_id/cloud/run_set_vcf_sample_id.py (hunk 1 above)
Comment (pgrosu) [inline set_vcf_sample_id/cloud/run_set_vcf_sample_id.py]: Wow, that is quite open-ended.  Maybe I'm misinterpreting, but do you mean that once a user is connected to a VM via the Pipeline API, that means that he/she can run any API without the need for credentials as those credentials are just propagated from the connection?  So then users have long-lived dynamic pipelines by running Apache servers as their `cmd`, which can accept bash scripts to process their results?

Comment (mbookman) [inline set_vcf_sample_id/cloud/run_set_vcf_sample_id.py]: A Compute Engine instance launched by the Pipelines API does not have end-user credentials on it.

By default, the instance uses the default Compute Engine service account associated with the Cloud project the instance is running in. For that instance, the service account is authorized to access resources within the Cloud project based on what [scopes](https://developers.google.com/identity/protocols/googlescopes) are specified. By default, the Cloud Storage, Genomics API, and Compute scopes are granted to instances launched by the Pipelines API.

A user is not "connected to a VM" for a pipeline (though the user might SSH to the VM to do some debugging), so no user credentials are "propagated".

Take a look at https://cloud.google.com/compute/docs/authentication for details on Compute Engine and service accounts.

Comment (pgrosu) [inline set_vcf_sample_id/cloud/run_set_vcf_sample_id.py]: This is helpful, but maybe I'm misreading something as I was hoping the service account would have pre-configured scopes.  So based on the following, a service account does not seem to have custom scopes attached to it automatically via the Console before creating an instance first - (I'm not referring to the default one):

https://cloud.google.com/compute/docs/authentication#using

https://cloud.google.com/storage/docs/authentication?hl=en#service_accounts

The idea being to be able to have a Builder Pipeline which will create and run a collection of pipelines, such as running dynamically-generated scripts to perform an ensemble analysis for picking the best analysis approach.  Thus the service account or private key that initiated the Builder Pipeline would have `full_control` scope for all APIs (i.e. Pipeline, Storage, Genomics, BigQuery, etc.) before the instances are dynamically created.  This way the authentication process would be transparent, and users just focus on doing the work and the credentials are automatic.  Can these APIs be enabled for domain-wide authority?

~p
